label-application-rule-config-service-impl/src/main/java/org/hypertrace/label/application/rule/config/service/LabelApplicationRuleValidatorImpl.java

```diff
@@ -0,0 +1,143 @@
+package org.hypertrace.label.application.rule.config.service;
+
+import org.hypertrace.core.grpcutils.context.RequestContext;
+import org.hypertrace.label.application.rule.config.service.v1.CreateLabelApplicationRuleRequest;
+import org.hypertrace.label.application.rule.config.service.v1.DeleteLabelApplicationRuleRequest;
+import org.hypertrace.label.application.rule.config.service.v1.GetLabelApplicationRuleRequest;
+import org.hypertrace.label.application.rule.config.service.v1.GetLabelApplicationRulesRequest;
+import org.hypertrace.label.application.rule.config.service.v1.LabelApplicationRuleData;
+import org.hypertrace.label.application.rule.config.service.v1.LabelApplicationRuleData.Action;
+import org.hypertrace.label.application.rule.config.service.v1.LabelApplicationRuleData.CompositeCondition;
+import org.hypertrace.label.application.rule.config.service.v1.LabelApplicationRuleData.JsonCondition;
+import org.hypertrace.label.application.rule.config.service.v1.LabelApplicationRuleData.LeafCondition;
+import org.hypertrace.label.application.rule.config.service.v1.LabelApplicationRuleData.StringCondition;
+import org.hypertrace.label.application.rule.config.service.v1.LabelApplicationRuleData.UnaryCondition;
+import org.hypertrace.label.application.rule.config.service.v1.UpdateLabelApplicationRuleRequest;
+
+public class LabelApplicationRuleValidatorImpl implements LabelApplicationRuleValidator {
+
+  @Override
+  public void validateOrThrow(
+      RequestContext requestContext,
+      CreateLabelApplicationRuleRequest createLabelApplicationRuleRequest) {
+    validateRequestContext(requestContext);
+    validateLabelApplicationRuleData(createLabelApplicationRuleRequest.getData());
+  }
+
+  @Override
+  public void validateOrThrow(
+      RequestContext requestContext,
+      GetLabelApplicationRuleRequest getLabelApplicationRuleRequest) {
+    validateRequestContext(requestContext);
+  }
+
+  @Override
+  public void validateOrThrow(
+      RequestContext requestContext,
+      GetLabelApplicationRulesRequest getLabelApplicationRulesRequest) {
+    validateRequestContext(requestContext);
+  }
+
+  @Override
+  public void validateOrThrow(
+      RequestContext requestContext,
+      UpdateLabelApplicationRuleRequest updateLabelApplicationRulesRequest) {
+    validateRequestContext(requestContext);
+    validateLabelApplicationRuleData(updateLabelApplicationRulesRequest.getData());
+  }
+
+  @Override
+  public void validateOrThrow(
+      RequestContext requestContext,
+      DeleteLabelApplicationRuleRequest deleteLabelApplicationRuleRequest) {
+    validateRequestContext(requestContext);
+  }
+
+  private void validateRequestContext(RequestContext requestContext) {
```

Comment:
You can use the common validation utils to replace this and many of the specific errors below.

Reply:
Getting this error when calling the common validation apis on some fields "Caused by: java.lang.UnsupportedOperationException: FieldDescriptor.getDefaultValue() called on an embedded message field".

Reply:
What was the corresponding code? I believe you'll hit that if you try to validate a message (as opposed to a primitive) is set, but that's by design.

Take a message like 
```proto
message WrapperMessage {
  InnerMessage inner_message = 1;
 
  message InnerMessage {
     string str_value = 1;
  }
}
```

If you want to make sure that str_value is set, you'd check it directly - and that should work. Checking if InnerMessage is set doesn't really provide validation value and I believe that's what would throw today. We could add support for it given a compelling use case, but I've yet to see one.

Reply:
Yes, you are right, but I think we still need to support it. I don't think we can validate a oneOf field using the given functions
Take this example 
```proto
message WrapperMessage {
  InnerMessage inner_message = 1;
 
  message InnerMessage {
        oneOf InnerMessage{
             string str_value = 1;
             int int_value = 2;
             ...
        }
  }
}
```
How do we validate it?

Reply:
for one of, we generally want to go in and validate whichever case, right?What my code looks like for that in other cases is for example:


```java

 private void validateEachOneOf(WrapperMessage wrapperMessage) {
    switch (wrapperMessage.getInnerMessageCase()) {
      case STR_VALUE:
        validateNonDefaultPresenceOrThrow(wrapperMessage, WrapperMessage.STR_VALUE_FIELD_NUMBER);
        break;
      case INT_VALUE:
        validateNonDefaultPresenceOrThrow(wrapperMessage, WrapperMessage.INT_VALUE_FIELD_NUMBER);
        break;
// ...
      default:
        throw Status.INVALID_ARGUMENT
            .withDescription("Unexpected case: " + printMessage(wrapperMessage))
            .asRuntimeException();
    }
  }
```

For the primitive cases, we could likely figure out a way to automate it, but I'm not sure we'd want to do implicit validation like that, since a change in proto _should_ correspond to either no change in validation logic (if that were to be a breaking change), or an explicit change (either in a new optional field case, such as one of, or a breaking change).

Reply:
Thanks, this works.